File: database/repositories/slots.py
from typing import Optional, Dict, Any, List

from database.client import supabase


FREE_STATUS = "Свободен"
WAITING_STATUS = "Ожидает подтверждения"
ACTIVE_STATUS = "Занят"


def get_available_slots(
    service_id: int,
    city: Optional[str] = None
) -> List[Dict[str, Any]]:
    query = (
        supabase
        .table("slots")
        .select("*")
        .eq("service_id", service_id)
        .eq("status", FREE_STATUS)
        .order("date_time")
    )

    if city:
        query = query.eq("city", city)

    response = query.execute()
    return response.data or []


def get_slot(slot_id: int) -> Optional[Dict[str, Any]]:
    response = (
        supabase
        .table("slots")
        .select("*")
        .eq("slot_id", slot_id)
        .limit(1)
        .execute()
    )

    if not response.data:
        return None

    return response.data[0]


def get_master_slots(master_id: int) -> List[Dict[str, Any]]:
    response = (
        supabase
        .table("slots")
        .select("*")
        .eq("master_id", master_id)
        .order("date_time")
        .execute()
    )

    return response.data or []


def get_master_active_slots(master_id: int) -> List[Dict[str, Any]]:
    response = (
        supabase
        .table("slots")
        .select("*")
        .eq("master_id", master_id)
        .in_("status", [
            FREE_STATUS,
            WAITING_STATUS,
            ACTIVE_STATUS,
        ])
        .order("date_time")
        .execute()
    )

    return response.data or []


def create_slot(
    master_id: int,
    service_id: int,
    date_time: str,
    city: str = "",
    address: str = "",
    description: str = "",
) -> Optional[Dict[str, Any]]:
    existing = (
        supabase
        .table("slots")
        .select("slot_id")
        .eq("master_id", master_id)
        .eq("date_time", date_time)
        .limit(1)
        .execute()
    )

    if existing.data:
        return None

    response = (
        supabase
        .table("slots")
        .insert({
            "master_id": master_id,
            "service_id": service_id,
            "date_time": date_time,
            "status": FREE_STATUS,
            "city": city,
            "address": address,
            "description": description,
        })
        .select()
        .single()
        .execute()
    )

    return response.data


def update_slot_status(
    slot_id: int,
    status: str
) -> Optional[Dict[str, Any]]:
    response = (
        supabase
        .table("slots")
        .update({"status": status})
        .eq("slot_id", slot_id)
        .select()
        .single()
        .execute()
    )

    return response.data


def cancel_slot(slot_id: int) -> Optional[Dict[str, Any]]:
    return update_slot_status(slot_id, "Отменен")

File: webapp/js/data/slots.js
import { supabase } from './supabase.js';


export async function getAvailableSlots(
    serviceId,
    city = null
) {
    if (!serviceId) {
        return [];
    }

    let query = supabase
        .from('slots')
        .select(`
            *,
            services (
                service_id,
                name,
                icon_url
            )
        `)
        .eq('service_id', serviceId)
        .eq('status', 'Свободен')
        .order('date_time');


    if (city && city !== 'Город') {
        query = query.eq(
            'city',
            city
        );
    }


    const { data, error } =
        await query;


    if (error) {
        throw error;
    }

    return data || [];
}


export async function getSlot(
    slotId
) {
    if (!slotId) {
        return null;
    }

    const { data, error } = await supabase
        .from('slots')
        .select(`
            *,
            services (
                service_id,
                name,
                icon_url
            )
        `)
        .eq('slot_id', slotId)
        .limit(1)
        .maybeSingle();

    if (error) {
        throw error;
    }

    return data;
}


export async function getMasterSlots(
    masterId
) {
    if (!masterId) {
        return [];
    }

    const { data, error } = await supabase
        .from('slots')
        .select(`
            *,
            services (
                service_id,
                name,
                icon_url
            )
        `)
        .eq('master_id', masterId)
        .order('date_time');

    if (error) {
        throw error;
    }

    return data || [];
}


export async function getMasterActiveSlots(
    masterId
) {
    if (!masterId) {
        return [];
    }

    const { data, error } = await supabase
        .from('slots')
        .select(`
            *,
            services (
                service_id,
                name,
                icon_url
            )
        `)
        .eq('master_id', masterId)
        .in(
            'status',
            [
                'Свободен',
                'Ожидает подтверждения',
                'Занят'
            ]
        )
        .order('date_time');

    if (error) {
        throw error;
    }

    return data || [];
}


export async function createSlot({
    masterId,
    serviceId,
    dateTime,
    city = '',
    address = '',
    description = ''
}) {
    if (!masterId) {
        throw new Error('Master ID is required.');
    }

    if (!serviceId) {
        throw new Error('Service ID is required.');
    }

    if (!dateTime) {
        throw new Error('Date and time are required.');
    }


    const { data: existingSlot, error: existingError } =
        await supabase
            .from('slots')
            .select('slot_id')
            .eq('master_id', masterId)
            .eq('date_time', dateTime)
            .limit(1)
            .maybeSingle();


    if (existingError) {
        throw existingError;
    }


    if (existingSlot) {
        throw new Error(
            'Слот на это время уже существует.'
        );
    }


    const { data, error } =
        await supabase
            .from('slots')
            .insert({
                master_id: masterId,
                service_id: serviceId,
                date_time: dateTime,
                status: 'Свободен',
                city,
                address,
                description
            })
            .select()
            .single();


    if (error) {
        throw error;
    }

    return data;
}


export async function updateSlotStatus(
    slotId,
    status
) {
    if (!slotId) {
        throw new Error('Slot ID is required.');
    }

    const { data, error } =
        await supabase
            .from('slots')
            .update({
                status
            })
            .eq('slot_id', slotId)
            .select()
            .single();

    if (error) {
        throw error;
    }

    return data;
}


export async function cancelSlot(
    slotId
) {
    return updateSlotStatus(
        slotId,
        'Отменен'
    );
}